### DISCUSSION TITLE
작은 따옴표 vs 큰 따옴표

### DISCUSSION THREAD

**Discussion by cbscsm** [8percent/styleguide#33] [General]
파이썬에서는 두 종류의 따옴표(`'`, `"`)를 동일하게 취급하고 있습니다. PEP8에서도 특정 따옴표를 사용하라고는 하지 않고, 통일해서만 사용하라고 하고 있습니다. 현재 코드 대부분에서 작은따옴표(`'`)를 메인으로 사용하고 있는 것으로 파악했는데, 가이드에 명시가 되면 좋을 것 같아 discussion을 만듭니다.

- [1안] 작은따옴표를 메인으로 사용. 본문에 작은따옴표가 필요할 때에는 큰따옴표 사용. 본문에 두 가지 따옴표가 모두 사용되는 경우 작은 따옴표와 이스케이프 사용.
  + 이 방식은 `str.__repr__`에서도 사용하는 방식입니다. 프로토콜상 `repr`의 표현 형태는 가능한경우 객체의 생성방법을 나타내기 때문에, 같은 방식으로 객체를 생성하는 것이 일관성이 있다고 볼 수 있습니다.
  + shift키를 적게 눌러도 되며, 폰트에 따라 큰따옴표 대비 가독성이 좋은 경우가 있습니다.
```python
'String'
"It's a string"
'"String"'
'"It\'s a string"'
```

- [1-1안] 항상 작은따옴표 사용. 내부에서 작은따옴표가 필요한 경우 이스케이프.
  + 위 방식과 다른점은, 큰따옴표를 사용하는 대신 이스케이프를 사용한다는 점입니다.
  + 이 부분때문에 `str.__repr__`의 구현과 달라집니다.
```python
'String'
'It\'s a string'
'"String"'
'"It\'s a string"'
```

- [2안] 항상 큰따옴표 사용. 내부에서 큰따옴표가 필요한 경우 이스케이프.
  + 이 방식은 다른 언어(두 따옴표를 모두 지원하지 않는 언어)와의 호환성이 좋다는 장점이 있습니다. 복붙 시 유용합니다. 예) C/C++, Java, JSON 등
```python
"String"
"It's a string"
"\"String\""
"\"It\'s a string\""
```

- [3안] 키에는 작은따옴표, 출력문장에는 큰따옴표 사용. 혹은 단어/문장으로 구분.
  + 간혹 이 방식을 사용하는 경우도 있지만, PEP에서 권장하는 일관성에 위배되기도 하며 작은따옴표와 큰따옴표의 사용을 구분하는 방식이 모호한 상황이 있다는 문제가 있습니다.
```python
my_dict['key']
print("Hello, world")
```

**Comment by cbscsm** [comment]
저는 1안을 선호합니다. 따옴표는 어떻게 쓰더라도 워낙 취향에 의존하는 부분이라 생각하는데, 파이썬 내부구현과 일치하는 표현법이기 때문에 조금이나마 합의를 이루는 기준이 있다는 점, 그리고 구분의 모호함도 없다는 점 때문입니다.

**Comment by Kirade** [comment]
저도 1안에 동의합니다.

**Comment by kyungjunleeme** [comment]
저도 1안에 동의합니다.

**Comment by leehosung** [comment]
왠지 불편한 마음이 들지만 ㅋㅋ 1안에 동의합니다.

**Comment by twins1040** [comment]
Black 의 방식은 어떤것이죠? Black 도 기본 따옴표를 `'`로 하는 것이 가능한 것으로 알고있습니다.
이때의 방식을 Django 가 따라갈 것이고 Python 도 같이 따라갔으면 해서요.

**Reply by twins1040** [reply]
1안이 Black 과도 같네요.
